enhancements/network/ovs-hardware-offload.md

```diff
@@ -0,0 +1,242 @@
+daemon and SR-IOV hardware NIC.
+
+- OpenvSwitch daemon is managed by Cluster Network Operator (CNO) and runs in
+`ovs-node` DaemonSet. A boolean configurable field will be added in CNO
+`Network` Custom Resource Definition `spec.defaultNetwork.ovnKubernetesConfig`
+to enable or disable OVS Hardware Offload feature for OpenvSwitch daemon.
+
+- SR-IOV hardware NIC is managed by SR-IOV Network Operator and will be
+provisioned with a new device type `switchdev`. The initial supported NIC
+model will be Mellanox MT27800 Family [ConnectX-5] 25GbE dual-port SFP28.
+
+Besides CNO and SR-IOV Operator changes, there are several other dependencies.
+
+- OVS Hardware Offload will only be supported on bare metal deployment.
+
+- Multus supports overwriting cluster network definition by specifying a
+customized default net-attach-def in pod annotation. Customized net-attach-def
+will still use `ovn-k8s-cni-overlay` CNI plugin, but with an special annotation
+that requests a SR-IOV device.
+
+- RHCOS image shall be based on RHEL-8.3 for fully support of OVS Hardware
+Offload with connection track feature.
+
+- OpenvSwitch package version 2.13.
+
+- Mellanox SR-IOV VF-LAG provides hardware link-aggregation (LAG) for hardware
+offloaded ports. It allows offloading of OVS rules to a linux `bond` network
+device that combines two idential and hardware offload capable Physical
+Functions (PFs). In order to use SR-IOV VF-LAG, a bond interface needs to be
+created and used as geneve endpoint of OVS bridge. With the SR-IOV VF LAG
+capability, the NIC PFs can receive the rules that the OVS tries to offload to
+the linux bond net device and offload them to the hardware e-switch.
+
+#### RHCOS
+
+- RHEL 8.3 be used as RHCOS image base for fully support of OVS Hardware Offload
+with connection track feature.
+
+#### OpenvSwitch
+
+- OpenvSwitch package [version 2.13](https://github.com/openshift/ovn-kubernetes/pull/122) is currently used in OpenShift.
+
+#### OVN-Kubernetes
+
+- OVS Hardware Offload is a [feature available](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/ovs_offload.md) in OVN-Kubernetes upstream.
+
+- `ovn-k8s-cni-overlay` CNI plugin configures the pod interface by moving the VF
+to pod namespace and plugging VF representor to `br-int` OVS bridge. The
+relevant [patches](https://github.com/openshift/ovn-kubernetes/commit/6c96467d0d3e58cab05641293d1c1b75e5914795) are available downstream.
+
+#### Cluster Network Operator
+
+- Cluster Network Operator will enable hardware offload and configure tc-policy
+on node OpenvSwitch instance and restart OpenvSwitch instance for hardware
+offload change to take effect.
+
+- Enabling `hw-offload` in CNO `ovs-node` DaemonSet is equivalent to executing
+`ovs-vsctl set Open_vSwitch . other_config:hw-offload=true`. The default value
+is `false`.
+
+- Setting `tc-policy` in CNO `ovs-node` DaemonSet is equivalent to executing
+`ovs-vsctl set Open_vSwitch . other_config:tc-policy=skip_sw`. The default
```

Comment:
The default value `none` uses both - writes to software and if available to hardware. Any particular reason to use `skip-sw`?

Reply:
no particuar reason, this is an example of `tc-policy`, it will be configurable by cluster admin.

Reply:
ovs hw offload is experimental in OVS project and is not in GA support on RHEL, until then I think we may want to have it disabled by default or be controlled per cluster. 

If `tc-policy` is configured to allow both sw and hw, the traffic can fall back to use ovs software datapath if flow cannot be offloaded to hw.

Reply:
@krsacme Using skip-sw allows ovs to switch back to ovs kernel datapath(the original datapath used on OpenShift) from hw offload failure, this excludes tc software datapath. There are uncertainties of using tc software datapath: is it matured enough? what's the performance look like compared with ovs kernel datapath?  what is the debuggbility?

Reply:
> There are uncertainties of using tc software datapath

I don't know about uncertainties on using tc. But it would recommeded to use the configuration which offload is validated with. OvS goes with the default options as tc than the ovs. Mellanox can be in a better possition to answer it.

> what's the performance look like compared with ovs kernel datapath? 

There is no analysis that i know of, between "none" and "skip_sw". It is a better to do that to choose the mode.

Reply:
> > There are uncertainties of using tc software datapath
> 
> I don't know about uncertainties on using tc. But it would recommeded to use the configuration which offload is validated with. OvS goes with the default options as tc than the ovs. Mellanox can be in a better possition to answer it.

Would the kernel be the default datapath type if not specified? or is it changed to use tc as default, in which ovs version?

Reply:
>But it would recommeded to use the configuration which offload is validated with. OvS goes with the default options as tc than the ovs. Mellanox can be in a better possition to answer it.

when `hw-offload` is enabled the default datapath would be TC, we recommend to use `tc-policy=None` which is how this feature is validated when SR-IOV is used.

Reply:
Is there any known issue when using tc-policy=skip-sw or it's just not validated?

Reply:
> Is there any known issue when using tc-policy=skip-sw or it's just not validated?

its not validated